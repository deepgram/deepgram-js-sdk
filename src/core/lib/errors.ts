export class DeepgramError extends Error {
  protected __dgError = true;

  constructor(message: string) {
    super(message);
    this.name = "DeepgramError";
  }
}

export function isDeepgramError(error: unknown): error is DeepgramError {
  return typeof error === "object" && error !== null && "__dgError" in error;
}

export class DeepgramApiError extends DeepgramError {
  status: number;

  constructor(message: string, status: number) {
    super(message);
    this.name = "DeepgramApiError";
    this.status = status;
  }

  toJSON() {
    return {
      name: this.name,
      message: this.message,
      status: this.status,
    };
  }
}

export class DeepgramUnknownError extends DeepgramError {
  originalError: unknown;

  constructor(message: string, originalError: unknown) {
    super(message);
    this.name = "DeepgramUnknownError";
    this.originalError = originalError;
  }
}

export class DeepgramVersionError extends DeepgramError {
  constructor() {
    super(
      `You are attempting to use an old format for a newer SDK version. Read more here: https://dpgr.am/js-v3`
    );

    this.name = "DeepgramVersionError";
  }
}

/**
 * Enhanced WebSocket error that captures additional debugging information
 * including status codes, request IDs, and response headers when available.
 */
export class DeepgramWebSocketError extends DeepgramError {
  originalEvent?: ErrorEvent | Event;
  statusCode?: number;
  requestId?: string;
  responseHeaders?: Record<string, string>;
  url?: string;
  readyState?: number;

  constructor(
    message: string,
    options: {
      originalEvent?: ErrorEvent | Event;
      statusCode?: number;
      requestId?: string;
      responseHeaders?: Record<string, string>;
      url?: string;
      readyState?: number;
    } = {}
  ) {
    super(message);
    this.name = "DeepgramWebSocketError";
    this.originalEvent = options.originalEvent;
    this.statusCode = options.statusCode;
    this.requestId = options.requestId;
    this.responseHeaders = options.responseHeaders;
    this.url = options.url;
    this.readyState = options.readyState;
  }

  toJSON() {
    return {
      name: this.name,
      message: this.message,
      statusCode: this.statusCode,
      requestId: this.requestId,
      responseHeaders: this.responseHeaders,
      url: this.url,
      readyState: this.readyState,
      originalEvent: this.originalEvent
        ? {
            type: this.originalEvent.type,
            timeStamp: this.originalEvent.timeStamp,
          }
        : undefined,
    };
  }
}
