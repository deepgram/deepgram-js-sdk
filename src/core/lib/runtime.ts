export const NODE_VERSION =
  typeof process !== "undefined" && process.versions && process.versions.node
    ? process.versions.node
    : "unknown";

export const BUN_VERSION =
  typeof process !== "undefined" && process.versions && process.versions.bun
    ? process.versions.bun
    : "unknown";

export const BROWSER_AGENT =
  typeof window !== "undefined" && window.navigator && window.navigator.userAgent
    ? window.navigator.userAgent
    : "unknown";

export const isBrowser = () => BROWSER_AGENT !== "unknown";

export const isNode = () => NODE_VERSION !== "unknown";

export const isBun = () => BUN_VERSION !== "unknown";
