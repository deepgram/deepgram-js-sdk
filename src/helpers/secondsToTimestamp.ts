export function secondsToTimestamp(seconds: number): string {
  return new Date(seconds * 1000).toISOString().substr(11, 12);
}
