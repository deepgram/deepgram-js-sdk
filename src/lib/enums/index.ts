export { LiveConnectionState } from "./LiveConnectionState";
export { LiveTranscriptionEvents } from "./LiveTranscriptionEvents";
