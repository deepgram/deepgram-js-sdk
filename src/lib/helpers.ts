import { Headers as CrossFetchHeaders } from "cross-fetch";
import {
  DeepgramClientOptions,
  FileSource,
  PrerecordedSource,
  UrlSource,
  TextSource,
  AnalyzeSource,
} from "./types";
import { Readable } from "stream";
import merge from "deepmerge";

export function stripTrailingSlash(url: string): string {
  return url.replace(/\/$/, "");
}

export const isBrowser = () => typeof window !== "undefined";
export const isServer = () => typeof process !== "undefined";

export function applySettingDefaults(
  options: DeepgramClientOptions,
  defaults: DeepgramClientOptions
): DeepgramClientOptions {
  return merge(defaults, options);
}

export function appendSearchParams(
  searchParams: URLSearchParams,
  options: Record<string, unknown>
): void {
  Object.keys(options).forEach((i) => {
    if (Array.isArray(options[i])) {
      const arrayParams = options[i] as Array<any>;
      arrayParams.forEach((param) => {
        searchParams.append(i, String(param));
      });
    } else {
      searchParams.append(i, String(options[i]));
    }
  });
}

export const resolveHeadersConstructor = () => {
  if (typeof Headers === "undefined") {
    return CrossFetchHeaders;
  }

  return Headers;
};

export const isUrlSource = (
  providedSource: PrerecordedSource | AnalyzeSource
): providedSource is UrlSource => {
  if ((providedSource as UrlSource).url) return true;

  return false;
};

export const isTextSource = (
  providedSource: PrerecordedSource | AnalyzeSource
): providedSource is TextSource => {
  if ((providedSource as TextSource).text) return true;

  return false;
};

export const isFileSource = (providedSource: PrerecordedSource): providedSource is FileSource => {
  if (isReadStreamSource(providedSource) || isBufferSource(providedSource)) return true;

  return false;
};

const isBufferSource = (providedSource: PrerecordedSource): providedSource is Buffer => {
  if (providedSource as Buffer) return true;

  return false;
};

const isReadStreamSource = (providedSource: PrerecordedSource): providedSource is Readable => {
  if (providedSource as Readable) return true;

  return false;
};

export class CallbackUrl extends URL {
  private callbackUrl = true;
}
