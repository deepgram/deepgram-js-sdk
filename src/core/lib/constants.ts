import { convertProtocolToWs } from "./helpers";
import { isBrowser, isBun, isNode, NODE_VERSION, BUN_VERSION, BROWSER_AGENT } from "./runtime";
import { version } from "./version";
import type { DefaultNamespaceOptions, DefaultClientOptions } from "./types";

const getAgent = () => {
  if (isNode()) {
    return `node/${NODE_VERSION}`;
  } else if (isBun()) {
    return `bun/${BUN_VERSION}`;
  } else if (isBrowser()) {
    return `javascript ${BROWSER_AGENT}`;
  } else {
    return `unknown`;
  }
};

export const DEFAULT_HEADERS = {
  "Content-Type": `application/json`,
  "X-Client-Info": `@deepgram/sdk; ${isBrowser() ? "browser" : "server"}; v${version}`,
  "User-Agent": `@deepgram/sdk/${version} ${getAgent()}`,
};

export const DEFAULT_URL = "https://api.deepgram.com";
export const DEFAULT_AGENT_URL = "wss://agent.deepgram.com";

export const DEFAULT_GLOBAL_OPTIONS: Partial<DefaultNamespaceOptions> = {
  fetch: { options: { url: DEFAULT_URL, headers: DEFAULT_HEADERS } },
  websocket: {
    options: { url: convertProtocolToWs(DEFAULT_URL), _nodeOnlyHeaders: DEFAULT_HEADERS },
  },
};

export const DEFAULT_AGENT_OPTIONS: Partial<DefaultNamespaceOptions> = {
  fetch: { options: { url: DEFAULT_URL, headers: DEFAULT_HEADERS } },
  websocket: {
    options: { url: DEFAULT_AGENT_URL, _nodeOnlyHeaders: DEFAULT_HEADERS },
  },
};

export const DEFAULT_OPTIONS: DefaultClientOptions = {
  global: DEFAULT_GLOBAL_OPTIONS,
  agent: DEFAULT_AGENT_OPTIONS,
};

export enum SOCKET_STATES {
  connecting = 0,
  open = 1,
  closing = 2,
  closed = 3,
}

export enum CONNECTION_STATE {
  Connecting = "connecting",
  Open = "open",
  Closing = "closing",
  Closed = "closed",
}
