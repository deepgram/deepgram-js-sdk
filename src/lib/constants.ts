import { isBrowser } from "./helpers";
import { FetchOptions } from "./types/Fetch";
import { version } from "./version";

export const DEFAULT_HEADERS = {
  "Content-Type": `application/json`,
  "X-Client-Info": `@deepgram/sdk; ${isBrowser() ? "browser" : "server"}; v${version}`,
  "User-Agent": `@deepgram/sdk/${version}`,
};

export const DEFAULT_URL = "https://api.deepgram.com";

export const DEFAULT_GLOBAL_OPTIONS = {
  url: DEFAULT_URL,
};

export const DEFAULT_FETCH_OPTIONS: FetchOptions = {
  headers: DEFAULT_HEADERS,
};

export const DEFAULT_OPTIONS = {
  global: DEFAULT_GLOBAL_OPTIONS,
  fetch: DEFAULT_FETCH_OPTIONS,
};
