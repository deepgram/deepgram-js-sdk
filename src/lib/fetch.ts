import crossFetch from "cross-fetch";
import { resolveHeadersConstructor } from "./helpers";
import type { Fetch } from "./types/Fetch";

export const resolveFetch = (): Fetch => {
  let _fetch: Fetch;
  if (typeof fetch === "undefined") {
    _fetch = crossFetch as unknown as Fetch;
  } else {
    _fetch = fetch;
  }
  return (...args) => _fetch(...args);
};

export const fetchWithAuth = (apiKey: string): Fetch => {
  const fetch = resolveFetch();
  const HeadersConstructor = resolveHeadersConstructor();

  return async (input, init) => {
    let headers = new HeadersConstructor(init?.headers);

    if (!headers.has("Authorization")) {
      headers.set("Authorization", `Token ${apiKey}`);
    }

    return fetch(input, { ...init, headers });
  };
};

export const resolveResponse = async () => {
  if (typeof Response === "undefined") {
    return (await import("cross-fetch")).Response;
  }

  return Response;
};
