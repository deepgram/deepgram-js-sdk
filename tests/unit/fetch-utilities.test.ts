import { resolveHeadersConstructor } from "../../src/lib/helpers";
import { resolveFetch, fetchWithAuth, resolveResponse } from "../../src/lib/fetch";

describe("Unit Tests - Fetch Utilities", () => {
  describe("resolveHeadersConstructor", () => {
    it("should return a valid Headers constructor", () => {
      const HeadersConstructor = resolveHeadersConstructor();

      expect(typeof HeadersConstructor).toBe("function");

      // Should be able to create a Headers instance
      const headers = new HeadersConstructor();
      expect(headers).toBeDefined();

      // Should be able to set and get headers
      headers.set("Content-Type", "application/json");
      expect(headers.get("Content-Type")).toBe("application/json");
    });
  });

  describe("resolveFetch", () => {
    it("should return the custom fetch when provided", () => {
      const customFetch = jest.fn();
      const resolved = resolveFetch(customFetch);

      expect(typeof resolved).toBe("function");

      // Should use the custom fetch when called
      resolved("test-url", {});
      expect(customFetch).toHaveBeenCalledWith("test-url", {});
    });

    it("should return a working fetch function when no custom fetch provided", () => {
      const resolved = resolveFetch();

      expect(typeof resolved).toBe("function");
      // Function should be callable (we can't easily test actual network calls in unit tests)
    });

    it("should handle fetch arguments correctly", () => {
      const customFetch = jest.fn().mockResolvedValue({ ok: true });
      const resolved = resolveFetch(customFetch);

      const url = "https://api.example.com/test";
      const options = { method: "POST", body: "test data" };

      resolved(url, options);

      expect(customFetch).toHaveBeenCalledWith(url, options);
    });
  });

  describe("fetchWithAuth", () => {
    it("should add Authorization header when not present", async () => {
      const mockFetch = jest.fn().mockResolvedValue({ ok: true });
      const apiKey = "test-api-key";
      const authenticatedFetch = fetchWithAuth(apiKey, mockFetch);

      await authenticatedFetch("https://api.example.com/test", {});

      expect(mockFetch).toHaveBeenCalledWith(
        "https://api.example.com/test",
        expect.objectContaining({
          headers: expect.any(Object),
        })
      );

      const callArgs = mockFetch.mock.calls[0][1];
      const headers = callArgs.headers;
      expect(headers.get("Authorization")).toBe("Token test-api-key");
    });

    it("should not override existing Authorization header", async () => {
      const mockFetch = jest.fn().mockResolvedValue({ ok: true });
      const apiKey = "test-api-key";
      const authenticatedFetch = fetchWithAuth(apiKey, mockFetch);

      const existingHeaders = new Headers();
      existingHeaders.set("Authorization", "Bearer existing-token");

      await authenticatedFetch("https://api.example.com/test", {
        headers: existingHeaders,
      });

      const callArgs = mockFetch.mock.calls[0][1];
      const headers = callArgs.headers;
      expect(headers.get("Authorization")).toBe("Bearer existing-token");
    });

    it("should preserve other headers", async () => {
      const mockFetch = jest.fn().mockResolvedValue({ ok: true });
      const apiKey = "test-api-key";
      const authenticatedFetch = fetchWithAuth(apiKey, mockFetch);

      const headers = { "Content-Type": "application/json", "X-Custom": "value" };

      await authenticatedFetch("https://api.example.com/test", { headers });

      const callArgs = mockFetch.mock.calls[0][1];
      const resultHeaders = callArgs.headers;
      expect(resultHeaders.get("Content-Type")).toBe("application/json");
      expect(resultHeaders.get("X-Custom")).toBe("value");
      expect(resultHeaders.get("Authorization")).toBe("Token test-api-key");
    });
  });

  describe("resolveResponse", () => {
    it("should return a valid Response constructor", async () => {
      const ResponseConstructor = await resolveResponse();

      expect(typeof ResponseConstructor).toBe("function");
      expect(ResponseConstructor.name).toBe("Response");

      // Should be able to create a Response instance
      const response = new ResponseConstructor("test body", { status: 200 });
      expect(response).toBeDefined();
      expect(response.status).toBe(200);
    });
  });
});
