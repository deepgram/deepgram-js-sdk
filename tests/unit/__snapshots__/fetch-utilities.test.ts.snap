// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit Tests - Fetch Utilities fetchWithAuth should create authenticated fetch function 1`] = `
[
  {
    "input": {
      "apiKey": "test-api-key",
      "customFetch": undefined,
    },
    "output": {
      "isFunction": true,
    },
  },
  {
    "input": {
      "apiKey": "test-api-key",
      "customFetch": "[CustomFunction]",
    },
    "output": {
      "isFunction": true,
    },
  },
]
`;

exports[`Unit Tests - Fetch Utilities resolveFetch should resolve fetch function correctly 1`] = `
[
  {
    "input": undefined,
    "output": {
      "isFunction": true,
      "name": "[Anonymous]",
    },
  },
  {
    "input": "[CustomFunction]",
    "output": {
      "isFunction": true,
      "name": "[Anonymous]",
    },
  },
]
`;

exports[`Unit Tests - Fetch Utilities resolveHeadersConstructor should return appropriate Headers constructor 1`] = `
{
  "canInstantiate": true,
  "isFunction": true,
  "name": "Headers",
}
`;

exports[`Unit Tests - Fetch Utilities resolveResponse should resolve Response constructor 1`] = `
{
  "canInstantiate": true,
  "isFunction": true,
  "name": "Response",
}
`;
